components/brave_news/browser/feed_v2_builder.cc

```diff
@@ -1122,240 +1332,125 @@ void FeedV2Builder::NotifyUpdateCompleted() {
-  add_items(initial_block);
-
-  // Step 2: We always add an advertisment after the first block.
-  // https://docs.google.com/document/d/1bSVHunwmcHwyQTpa3ab4KRbGbgNQ3ym_GHvONnrBypg/edit#heading=h.82154jsxm16
-  auto advert = GenerateAd();
-  DVLOG(1) << "Step 2: Advertisement";
-  add_items(advert);
-
-  // Step 3: Generate a top news block
-  // https://docs.google.com/document/d/1bSVHunwmcHwyQTpa3ab4KRbGbgNQ3ym_GHvONnrBypg/edit#heading=h.7z05nb4b269d
-  // This block is a bit special - we take the top articles from the top few
-  // topics and display them in a cluster. If there are no topics, we try and do
-  // the same thing, but with the Top News channel.
-  auto top_news_block = GenerateTopTopicsBlock(publishers, topics);
-  if (top_news_block.empty()) {
-    top_news_block =
-        GenerateChannelBlock(locale, publishers, kTopNewsChannel, articles);
-  }
-  DVLOG(1) << "Step 3: Top News Block";
-  add_items(top_news_block);
-
-  // Repeat step 4 - 6 until we don't have any more articles to add to the feed.
-  constexpr uint8_t kIterationTypes = 3;
-  uint32_t iteration = 0;
-  while (true) {
-    std::vector<mojom::FeedItemV2Ptr> items;
-
-    auto iteration_type = iteration % kIterationTypes;
-
-    // Step 4: Block Generation
-    // https://docs.google.com/document/d/1bSVHunwmcHwyQTpa3ab4KRbGbgNQ3ym_GHvONnrBypg/edit#heading=h.os2ze8cesd8v
-    if (iteration_type == 0) {
-      DVLOG(1) << "Step 4: Standard Block";
-      items = GenerateBlockFromContentGroups(articles, locale, publishers,
-                                             eligible_content_groups);
-    } else if (iteration_type == 1) {
-      // Step 5: Block or Cluster Generation
-      // https://docs.google.com/document/d/1bSVHunwmcHwyQTpa3ab4KRbGbgNQ3ym_GHvONnrBypg/edit#heading=h.tpvsjkq0lzmy
-      // Half the time, a normal block
-      if (TossCoin()) {
-        DVLOG(1) << "Step 5: Standard Block";
-        items = GenerateBlockFromContentGroups(articles, locale, publishers,
-                                               eligible_content_groups);
-      } else {
-        items = GenerateClusterBlock(locale, publishers, subscribed_channels,
-                                     topics, articles);
-      }
-    } else if (iteration_type == 2) {
-      // Step 6: Optional special card
-      // https://docs.google.com/document/d/1bSVHunwmcHwyQTpa3ab4KRbGbgNQ3ym_GHvONnrBypg/edit#heading=h.n1ipt86esc34
-      if (TossCoin()) {
-        DVLOG(1) << "Step 6: Special Block";
-        items = GenerateSpecialBlock(suggested_publisher_ids);
-      } else {
-        DVLOG(1) << "Step 6: None (approximately half the time)";
-      }
-    } else {
-      NOTREACHED();
-    }
+void FeedV2Builder::GenerateFeed(UpdateSettings settings,
+                                 mojom::FeedV2TypePtr type,
+                                 FeedGenerator generator,
+                                 BuildFeedCallback callback) {
+  DCHECK_CALLED_ON_VALID_THREAD(thread_checker_);
```

Comment:
`thread_checker_` is invalid here. We are post-tasking to any thread in pool, so the different tasks could use different threads.
Let's just check !UI.